Add support for Docker networks.

bsideup opened the pull request:
Hi.

One of the highly requested features, especially from users who wanted to run a fleet of containers, i.e. Hazelcast (hey @gAmUssA :D )

Also very helpful for Selenium + container-under-the-test scenario

As you can see, I experimented a bit with DSL, and it seems very usable :)

martin-g commented:
It would be nice if some user documentation is added.

The signature of the following two is not very intuitive:
```
 public Network.JUnitRule network = newNetwork().as(Network.JUnitRule.class);
```
```
 Network network = newNetwork().as(Network.AutoCreated.class);
```

rnorth commented:
I'd second @martin-g's comment here; the `as` approach is clever but I worry that it's not intuitive. I think anybody using this would have to be copying-and-pasting from an example. I'm not sure that somebody using just method signatures in-IDE or Javadocs would be able to work out that they have to do this. 

Please can we re-think? Maybe something simpler (more coupled to JUnit) for the initial release, then go with a universal `as(JUnit4Rule.class)` pattern for both networks **and** containers in v2 (or similar)...

rnorth commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ All notable changes to this project will be documented in this file.
 - Added pre-flight checks (can be disabled with `checks.disable` configuration property) (#363)
 - Removed unused Jersey dependencies (#361)
 - Fixed non-POSIX fallback for file attribute reading (#371)
+- Added support for the Docker networks (#372)
```
Trivial grammar note - `the` is unnecessary here

bsideup replied:
Thanks! 👍 One day I will learn how to use them properly :D

asafm commented on core/src/main/java/org/testcontainers/containers/Network.java:
```diff
@@ -0,0 +1,107 @@
+    boolean isCreated();
+
+    default boolean create() {
+        return getId() != null;
```
I would leave out the default implementation- it's too magic: you need to guess that getId is doing the actual create. 
Regarding the Boolean - why not fail if can't create? In most chances you can't continue without a network.

bsideup replied:
default implementation here assumes that there is kinda no need to create a network. If id is known then what else to "create"?

The method returns boolean instead of an exception because it's up to the consumer how to react on that. However, I agree that `before` should throw if it returns false

bsideup replied:
Personally, I prefer the previous approach with `.as()` where the contract is clear, but for now we decided to go with an old approach until 2.0

asafm replied:
In the impl you placed the logic of creating the network in getId() - that was my point. In the default implementation you assumed getId will be called this create the network if needed. Am I missing something?

kiview replied:
I think @asafm point is, that `getId()` method here is definitly not a regular getter as one might expect, since it does the actual creation as well. So maybe it's having elegant code leveraging Lombok vs. self-explanatory code at this point.

bsideup replied:
`id` is the only thing needed to use this network. Default implementation assumes "no creation needed" semantic and returns true if id exists. **Do not** read the implementation to understand the idea of default implementation here.

Example:
```
class ExistingNetwork implements Network {
    String getId() {
        return "some-predefined-id";
    }
}
```

asafm commented on core/src/main/java/org/testcontainers/utility/ResourceReaper.java:
```diff
@@ -145,22 +142,34 @@ public void removeNetworks(String identifier) {
+                // First try to remove by name
+                dockerClient.removeNetworkCmd(networkName).exec();
+            } catch (Exception e) {
+                LOGGER.trace("Error encountered removing network by name ({}) - it may not have been removed", networkName);
```
Two tests running with same network name will fail , at least the second create will. They will not understand why since the failure is hidden under trace level. I would keep this at error level or warn.

bsideup replied:
network's name is random (see NetworkImpl)

asafm replied:
So no way of allowing me to specify a name? I'm Starting to build Testclusters, where I'm reusing containers across runs to shorten test duration. My assumption is same network name across runs. Would love to use the class created here for this.

kiview replied:
If your use case would reuse the containers, it would reuse the network as well, I assume?

bsideup replied:
@asafm we don't allow users to set names of containers (and now networks) because it might lead to the names conflict (especially with parallel execution). 

As @kiview said: if you reuse things, reuse the networks also

asafm replied:
I'm reusing across builds, on same machine. Quite similar to docker compose. I can code it with the client my self, but I think it's harmless  to leave the opening to set the name to who ever needs it. I'm ok with any direction you take.

rnorth replied:
Let's think about whether we could allow reuse between runs separately. It's probably quite doable, but for the sake of this PR let's resume later.

kiview commented on core/src/main/java/org/testcontainers/containers/Network.java:
```diff
@@ -0,0 +1,107 @@
+
+    String getName();
+
+    Boolean getEnableIpv6();
```
Do you really want a `Boolean`?

kiview commented on core/src/main/java/org/testcontainers/containers/Network.java:
```diff
@@ -0,0 +1,107 @@
+
+        @Getter(lazy = true)
+        private final String id = ((Supplier<String>) () -> {
+            ResourceReaper.instance().registerNetworkForCleanup(getName());
```
I'd prefer to have this in a private method.

kiview commented on core/src/main/java/org/testcontainers/containers/Network.java:
```diff
@@ -0,0 +1,107 @@
+
+        private final String name = UUID.randomUUID().toString();
+
+        private Boolean enableIpv6;
```
Same thing with the `Boolean` here.

bsideup replied:
null here means "the value wasn't provided and `withEnableIpv6` should not be called (use a default value from `CreateNetworkCmd`)

asafm commented on core/src/main/java/org/testcontainers/containers/Network.java:
```diff
@@ -0,0 +1,107 @@
+
+    @SneakyThrows
+    @SuppressWarnings("unchecked")
+    default <T extends Network> T as(Class<T> type) {
```
I'm curious to that purpose of this class, which is related to another question I had: why have an interface? Do we plan on users extending and supplying their own Network implementation?

asafm commented on core/src/main/java/org/testcontainers/containers/Network.java:
```diff
@@ -0,0 +1,107 @@
+        private Set<Consumer<CreateNetworkCmd>> createNetworkCmdModifiers = new LinkedHashSet<>();
+
+        @Getter(lazy = true)
+        private final String id = ((Supplier<String>) () -> {
```
I don't understand why a variable of type String receives a lambda function, casted to a Suplier of String and then executed? I mean - what's going on?

bsideup replied:
`lazy = true` is the answer

asafm commented on core/src/main/java/org/testcontainers/containers/Network.java:
```diff
@@ -0,0 +1,107 @@
+        @Override
+        public boolean isCreated() {
+            // Lombok with @Getter(lazy = true) will use AtomicReference as a field type for id
+            return ((AtomicReference<String>) (Object) id).get() != null;
```
Just a personal opinion - that's too much magic for such a simple thing as having a field called Id of type String. Maybe it will more clear if for this field, Lombok won't be used.

bsideup commented:
- Simplified the code
- Removed getters from `Network` interface (can be accessed with NetworkImpl if needed)
- Removed `isCreated`
- Less Lombok magic
